### PR TITLE
Fixes #3471 - prevent multiple marathon instances to fight over reser…

### LINKED ISSUE #3471: Resident Tasks: Prevent multiple marathon instances to fight over reservations
After implementing #3438, unknown/unneeded reservations will be garbage collected. If multiple Marathon instances are registered with the same Mesos role and start resident tasks, this leads to disaster.  All reservations for Marathon instance A are unknown to Marathon instance B and will be unreserved and destroy.

We can fix this by e.g. labeling the resources with the frameworkId and only unreserve/destroy resources with a matching frameworkId. The framework name does not work because Mesos allows multiple frameworks with the same name.

### PR DESCRIPTION
**Pull request by kolloch**
…vations

by also labeling the resources with the framework ID.

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package mesosphere.marathon.core.launcher.impl
+
+import org.scalatest.{GivenWhenThen, FunSuite}
+
+class TaskLabelsTest extends FunSuite with GivenWhenThen
```
Comment (meichstedt) [inline src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala]: That's a clean test ;)

**Comment by meichstedt** [comment]
Thanks, @kolloch – Only one comment: you added a `TaskLabelsTest` without any test cases. Maybe add some? Apart from that, LGTM!

File: src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala (hunk 1 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala]: Indeed. Amazingly clean!

**Comment by kolloch** [comment]
I didn't push my latest revision. Sorry.

**Comment by kolloch** [comment]
```diff
diff --git a/src/main/scala/mesosphere/marathon/core/launcher/impl/TaskLabels.scala b/src/main/scala/mesosphere/marathon/core/launcher/impl/TaskLabels.scala
index ef0c0cb..45aed48 100644
--- a/src/main/scala/mesosphere/marathon/core/launcher/impl/TaskLabels.scala
+++ b/src/main/scala/mesosphere/marathon/core/launcher/impl/TaskLabels.scala
@@ -11,6 +11,10 @@ object TaskLabels {
   private[this] final val FRAMEWORK_ID_LABEL = "marathon_framework_id"
   private[this] final val TASK_ID_LABEL = "marathon_task_id"

+  /**
+    * Returns a the task id for which this reservation has been performed if the reservation was
+    * labeled by this framework.
+    */
   def taskIdForResource(frameworkId: FrameworkId, resource: MesosProtos.Resource): Option[Task.Id] = {
     val labels = ResourceLabels(resource)
```

and the tests.

File: src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    taskIds should be(empty)
+  }
+
+  class Fixture {
+    import scala.collection.JavaConverters._
+
+    val appId = PathId("/test")
+    val taskId = Task.Id.forApp(appId)
+    val frameworkId = MarathonTestHelper.frameworkId
+    val otherFrameworkId = FrameworkId("very other different framework id")
```
Comment (meichstedt) [inline src/test/scala/mesosphere/marathon/core/launcher/impl/TaskLabelsTest.scala]: Nice test suite! :grin:

**Comment by meichstedt** [comment]
```
       /\/\

                       /\/\
        O  O  O
                O       Awesome!
               __|__
               || ||_____
               || ||    |
    --------------------------
     \   O   O   O   O      /
~ ~ ~ ~ ~ ~ ~ ~  ~ ~ ~ ~ ~ ~ ~ ~
 ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
```

**Pull request merged**